src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultAssemblyPartDiscoveryProvider.cs

```diff
@@ -37,7 +46,106 @@ public static IEnumerable<ApplicationPart> DiscoverAssemblyParts(string entryPoi
             var entryAssembly = Assembly.Load(new AssemblyName(entryPointAssemblyName));
             var context = DependencyContext.Load(entryAssembly);
 
-            return GetCandidateAssemblies(entryAssembly, context).Select(p => new AssemblyPart(p));
+            var candidateAssemblies = GetCandidateAssemblies(entryAssembly, context);
+            var additionalReferences = candidateAssemblies
+                .ToDictionary(
+                    ca => ca,
+                    ca => ca
+                        .GetCustomAttributes<AssemblyMetadataAttribute>()
+                        .Where(ama => ama.Key.Equals("Microsoft.AspNetCore.Mvc.AdditionalReference", StringComparison.Ordinal)).ToArray());
+
+            // Find out all the additional references defined by the assembly.
+            // [assembly: AssemblyMetadataAttribute("Microsoft.AspNetCore.Mvc.AdditionalReference", "Library.PrecompiledViews.dll,true|false")]
+            var additionalAssemblies = new List<AdditionalReference>();
+            foreach (var kvp in additionalReferences)
+            {
+                if (kvp.Value.Length > 0)
+                {
+                    foreach (var metadataAttribute in kvp.Value)
+                    {
+                        var fileName = metadataAttribute.Value.Substring(0, metadataAttribute.Value.IndexOf(","));
+                        var filePath = Path.Combine(Path.GetDirectoryName(kvp.Key.Location), fileName);
+                        var additionalReference = new AdditionalReference
+                        {
+                            File = filePath,
+                            IncludeByDefault = metadataAttribute.Value.Substring(metadataAttribute.Value.IndexOf(",") + 1).Equals("true")
```

Comment:
We  discussed some ideas for making this more first class and plan to revisit after preview1.